Add SAM2 point segmentation example

claydugo opened the pull request:
![resegmin](https://github.com/user-attachments/assets/0d9d06e3-5b73-4752-9aea-3747e39d9132)

This is a somewhat minimal example of how I interact with [SAM2](https://github.com/facebookresearch/sam2) using pygfx currently. 

I saw the proposed [scheduling changes](https://github.com/pygfx/wgpu-py/pull/618) and thought it might be helpful to have this ready for comparison.

claydugo commented:
Marking as draft because I would like to get it to be fast again after the scheduling changes.

hmaarrfk commented:
@claydugo i'm somewhat curious as to why any "qt import" needs to be made in this example. you might be able to get it fast now if you change the loop to be in a standard "pygfx managed" callback, and never import Qt whatsoever in your example.

Just a though.

hmaarrfk commented:
yeah it might be a good example of how the "non-async workflow might suffer"!

almarklein commented:
I did `pip install opencv-python sam2` and this alllowed me to run the example. It took me a minute to figure out that I had to move the green dot around :)  

I also tried with the scheduling branch and I also observed what you told yesterday: the fps going down to 1. It's stuck at exactly 1 because that's the default `min_fps`. The problem is that the canvas does not receive the request for a new draw. I fixed that in the scheduling branch, and will also apply that in the `rendercanvas.RenderCanvas`.

Some more context: the `qt.WgpuCanvas` is intended to be used as a toplevel window. It wraps the `qt.QWgpuWidget`. In `main` there's a few lines:
```py
def update(self):
        super().update()
        self._subwidget.update()
```
That were lost in the scheduling branch. I restored it now.

I actually expected people to use `canvas.request_draw()` and not `canvas.update()`, but obviously the latter makes much more sense once you're in qt land. So I also overloaded `update()` to call `request_draw()`.  The effect is that it plays nice with the scheduling mechanics, instead of letting qt decide when to draw.

Small tip when using a canvas in a Qt app: use `qt.QWgpuCanvas`  (or soon the `rendercanvas.qt.QRenderCanvas`). It safes a little indirection by not using the wrapper widget.

claydugo commented:
>I did pip install opencv-python sam2 and this alllowed me to run the example. It took me a minute to figure out that I had to move the green dot around :)

Haha sorry about that. I added some instructions to the docstring to hopefully clear that up. 

I also removed `opencv`, I think this is written in a way that a user can drop in a larger image to accomplish the same test. 

I modified this example to use `QRenderCavnas` now. For testing it required me to apply the following patch:

<details>
  <summary>Renderer rendercanvas patch</summary>

```diff
diff --git a/pygfx/renderers/wgpu/engine/renderer.py b/pygfx/renderers/wgpu/engine/renderer.py
index e11ce42..376f9a5 100644
--- a/pygfx/renderers/wgpu/engine/renderer.py
+++ b/pygfx/renderers/wgpu/engine/renderer.py
@@ -9,6 +9,7 @@ import weakref
 import numpy as np
 import wgpu
 import pylinalg as la
+import rendercanvas
 
 from ....objects._base import id_provider
 from ....objects import (
@@ -130,7 +131,15 @@ class WgpuRenderer(RootEventHandler, Renderer):
         super().__init__(*args, **kwargs)
 
         # Check and normalize inputs
-        if not isinstance(target, (Texture, GfxTextureView, wgpu.gui.WgpuCanvasBase)):
+        if not isinstance(
+            target,
+            (
+                Texture,
+                GfxTextureView,
+                wgpu.gui.WgpuCanvasBase,
+                rendercanvas.BaseRenderCanvas,
+            ),
+        ):
             raise TypeError(
                 f"Render target must be a Canvas or Texture, not a {target.__class__.__name__}"
             )
@@ -148,7 +157,7 @@ class WgpuRenderer(RootEventHandler, Renderer):
         # Get target format
         self.gamma_correction = gamma_correction
         self._gamma_correction_srgb = 1.0
-        if isinstance(target, wgpu.gui.WgpuCanvasBase):
+        if isinstance(target, (wgpu.gui.WgpuCanvasBase, rendercanvas.BaseRenderCanvas)):
             self._canvas_context = self._target.get_context()
             # Select output format. We currently don't have a way of knowing
             # what formats are available, so if not srgb, we gamma-correct in shader.
@@ -218,7 +227,9 @@ class WgpuRenderer(RootEventHandler, Renderer):
         """
         if self._pixel_ratio is not None:
             return self._pixel_ratio
-        elif isinstance(self._target, wgpu.gui.WgpuCanvasBase):
+        elif isinstance(
+            self._target, (wgpu.gui.WgpuCanvasBase, rendercanvas.BaseRenderCanvas)
+        ):
             target_pixel_ratio = self._target.get_pixel_ratio()
             if target_pixel_ratio > 1.0:
                 return target_pixel_ratio
@@ -269,7 +280,7 @@ class WgpuRenderer(RootEventHandler, Renderer):
     def logical_size(self):
         """The size of the render target in logical pixels."""
         target = self._target
-        if isinstance(target, wgpu.gui.WgpuCanvasBase):
+        if isinstance(target, (wgpu.gui.WgpuCanvasBase, rendercanvas.BaseRenderCanvas)):
             return target.get_logical_size()
         elif isinstance(target, Texture):
             return target.size[:2]  # assuming pixel-ratio 1
@@ -334,7 +345,9 @@ class WgpuRenderer(RootEventHandler, Renderer):
         # If the blend mode has changed, we may need a new _wobject_pipelines
         self._set_wobject_pipelines()
         # If our target is a canvas, request a new draw
-        if isinstance(self._target, wgpu.gui.WgpuCanvasBase):
+        if isinstance(
+            self._target, (wgpu.gui.WgpuCanvasBase, rendercanvas.BaseRenderCanvas)
+        ):
             self._target.request_draw()
 
     @property
@@ -361,7 +374,9 @@ class WgpuRenderer(RootEventHandler, Renderer):
     @gamma_correction.setter
     def gamma_correction(self, value):
         self._gamma_correction = 1.0 if value is None else float(value)
-        if isinstance(self._target, wgpu.gui.WgpuCanvasBase):
+        if isinstance(
+            self._target, (wgpu.gui.WgpuCanvasBase, rendercanvas.BaseRenderCanvas)
+        ):
             self._target.request_draw()
 
     def _set_wobject_pipelines(self):
@@ -569,7 +584,7 @@ class WgpuRenderer(RootEventHandler, Renderer):
             target = self._target
 
         # Get the wgpu texture view.
-        if isinstance(target, wgpu.gui.WgpuCanvasBase):
+        if isinstance(target, (wgpu.gui.WgpuCanvasBase, rendercanvas.BaseRenderCanvas)):
             wgpu_tex_view = self._canvas_context.get_current_texture().create_view()
         elif isinstance(target, Texture):
             need_mipmaps = target.generate_mipmaps
```

</details>

It is now running smoothly again on the `scheduling` branch. Thanks for the thorough review!

zetyquickly commented:
It would be cool to use WebGPU ONNX Runtime instead of PyTorch backend 
https://onnxruntime.ai/docs/tutorials/web/ep-webgpu.html

claydugo commented:
This is ready for review!

hmaarrfk commented on examples/other/sam2_point_segmentation.py:
```diff
@@ -0,0 +1,256 @@
+This script demonstrates a real-time interactive segmentation application using pygfx points as prompts to the SAM2 (Segment Anything Model 2) model and Qt (PySide6) for the GUI.
+
+Additional dependencies required to run this example:
+    pip install git+https://github.com/facebookresearch/sam2.git
```
Thank you. 

Just a note I suggested that we use this strategy since the pypi package is not from the official distribution and quite old. 

Mostly the fact that it is highly not official and undocumented scares me.

almarklein reviewed (commented):
Only one small comment.
